src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceDescriptor.cs

```diff
@@ -58,25 +89,51 @@ public class ServiceDescriptor
             Type serviceType,
             Func<IServiceProvider, object> factory,
             ServiceLifetime lifetime)
-            : this(serviceType, lifetime)
+            : this(serviceType, serviceKey: null, lifetime)
         {
             ThrowHelper.ThrowIfNull(serviceType);
             ThrowHelper.ThrowIfNull(factory);
 
             ImplementationFactory = factory;
         }
 
-        private ServiceDescriptor(Type serviceType, ServiceLifetime lifetime)
+        /// <summary>
+        /// Initializes a new instance of <see cref="ServiceDescriptor"/> with the specified <paramref name="factory"/>.
+        /// </summary>
+        /// <param name="serviceType">The <see cref="Type"/> of the service.</param>
+        /// <param name="serviceKey">The <see cref="ServiceDescriptor.ServiceKey"/> of the service.</param>
+        /// <param name="factory">A factory used for creating service instances.</param>
+        /// <param name="lifetime">The <see cref="ServiceLifetime"/> of the service.</param>
+        public ServiceDescriptor(
+            Type serviceType,
+            object? serviceKey,
+            Func<IServiceProvider, object, object> factory,
+            ServiceLifetime lifetime)
+            : this(serviceType, serviceKey, lifetime)
+        {
+            ThrowHelper.ThrowIfNull(serviceType);
+            ThrowHelper.ThrowIfNull(factory);
+
+            KeyedImplementationFactory = factory;
+        }
+
+        private ServiceDescriptor(Type serviceType, object? serviceKey, ServiceLifetime lifetime)
         {
             Lifetime = lifetime;
             ServiceType = serviceType;
+            ServiceKey = serviceKey;
         }
 
         /// <summary>
         /// Gets the <see cref="ServiceLifetime"/> of the service.
         /// </summary>
         public ServiceLifetime Lifetime { get; }
 
+        /// <summary>
+        /// Get the key of the service, if applicable.
+        /// </summary>
+        public object? ServiceKey { get; }
```

Comment:
If the developper calls `GetKeyedService` with an unsopported DI provider, it will throw already.

But currently, it will fail "silently" when using the attributes, which is not ideal. Maybe we could check when adding to the `IServiceCollection`, by checking if it contains some marker?

Reply:
Whether or not the `IServiceCollection` contains some marker would not be enough to indicate which container will be used to build the `IServiceCollection`. The validation must come later.

And the problem is necessarily just if someone calls `GetKeyedService` or if someone uses the `FromKeyedServicesAttribute`. It's also a problem if some library registers a keyed service that no one ever intended to resolve, but now it overrides a normal service in containers that don't know about keyed services.

Reply:
Duplicating all property with `Keyed` in the name like I did for `KeyedImplementationFactory` is definitely doable but doesn't feel right to me. 

Ideally it should throw when building the container (or better when adding it to the collection, but that might not be doable)